cli/command/container/utils.go

```diff
@@ -13,28 +14,74 @@ import (
 	"golang.org/x/net/context"
 )
 
-func waitExitOrRemoved(ctx context.Context, dockerCli *command.DockerCli, containerID string, waitRemove bool) chan int {
+func waitExitOrRemoved(ctx context.Context, dockerCli *command.DockerCli, containerID string, attachDone chan struct{}, waitRemove bool) chan int {
 	if len(containerID) == 0 {
 		// containerID can never be empty
 		panic("Internal Error: waitExitOrRemoved needs a containerID as parameter")
 	}
 
-	var removeErr error
 	statusChan := make(chan int)
-	exitCode := 125
+	ctxWithCancel, cancel := context.WithCancel(ctx)
 
-	// Get events via Events API
 	f := filters.NewArgs()
 	f.Add("type", "container")
 	f.Add("container", containerID)
 	options := types.EventsOptions{
 		Filters: f,
 	}
+
+	go func() {
+		statusChan <- waitExitedOrRemovedEvent(ctxWithCancel, dockerCli, containerID, waitRemove, options)
+		cancel()
+	}()
+
+	if waitRemove {
+		// This is a fallback in case there are issues with the events API.
+		// Uses polling at pretty long intervals instead of events
+		// This makes sure we don't hang the CLI.
+		go func() {
+			select {
+			case <-attachDone:
+			case <-ctxWithCancel.Done():
+				return
+			}
+
+			waitRemovedPolling(ctxWithCancel, dockerCli, containerID)
+			cancel()
+
+			// try to get a replay of events so we can collect the exit status
+			// but still timeout in case of missed events
+			ctxWithTimeout, cancel := context.WithTimeout(ctx, 30*time.Second)
+			options.Since = "0"
+			statusChan <- waitExitedOrRemovedEvent(ctxWithTimeout, dockerCli, containerID, false, options)
+			cancel()
+		}()
+	}
+
+	return statusChan
+}
+
+func waitRemovedPolling(ctx context.Context, dockerCli *command.DockerCli, containerID string) {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			_, err := dockerCli.Client().ContainerInspect(ctx, containerID)
```

Comment:
The `waitRemovedPolling` rely on line 34: https://github.com/docker/docker/pull/31744/files#diff-1fe5ea7c675b9700809b89d15a414576R34 to cancel the context right? and line 34 `waitExitedOrRemovedEvent` has a param of `ctxWithCancel` which doesn't have a timeout, so it still rely on Event API.  If container fail to remove, there's no `destroy` event and nobody will cancel the context. 30s timeout only works after `waitRemovedPolling` returns.

That's why I think it removal failed, this could still hang~